src/ripple/nodestore/impl/DeterministicShard.cpp

```diff
@@ -0,0 +1,241 @@
+
+    auto digest = [&](int n) {
+        auto const data{hash.data()};
+        std::uint64_t result{0};
+
+        switch (n)
+        {
+            case 0:
+            case 1:
+                // Construct 64 bits from sequential eight bytes
+                for (int i = 0; i < 8; i++)
+                    result = (result << 8) + data[n * 8 + i];
+                break;
+
+            case 2:
+                // Construct 64 bits using the last four bytes of data
+                result = (static_cast<std::uint64_t>(data[16]) << 24) +
+                    (static_cast<std::uint64_t>(data[17]) << 16) +
+                    (static_cast<std::uint64_t>(data[18]) << 8) +
+                    (static_cast<std::uint64_t>(data[19]));
+                break;
+        }
+
+        return result;
+    };
+    auto const uid{digest(0)};
+    auto const salt{digest(1)};
+    auto const appType{digest(2) | deterministicType};
+
+    maxMemObjs_ = db->ledgersPerShard() <= 256 ? maxMemObjsTest : maxMemObjsDef;
+
+    // Open or create the NuDB key/value store
+    try
+    {
+        if (exists(tempdir_))
+            remove_all(tempdir_);
+
+        backend_->open(true, appType, uid, salt);
+    }
+    catch (std::exception const& e)
+    {
+        backend_.reset();
+        remove_all(tempdir_);
+        JLOG(j_.fatal()) << "deterministic shard " << index_
+                         << " failed to create due to exception: " << e.what()
+                         << ". Caught in function " << __func__;
+        return false;
+    }
+
+    return true;
+}
+
+std::shared_ptr<DeterministicShard>
+make_DeterministicShard(
+    Application& app,
+    boost::filesystem::path const& shardDir,
+    std::uint32_t index,
+    Serializer const& finalKey,
+    beast::Journal j)
+{
+    std::shared_ptr<DeterministicShard> detShard(
+        new DeterministicShard(app, shardDir, index, j));
+    if (!detShard->init(finalKey))
+        return {};
+    return detShard;
+}
+
+void
+DeterministicShard::close(bool cancel)
+{
+    if (!backend_)
+        return;
+
+    try
+    {
+        if (cancel)
+        {
+            backend_.reset();
+            remove_all(tempdir_);
+        }
+        else
+        {
+            flush(true);
+            backend_.reset();
+            remove_all(finaldir_);
+            rename(tempdir_, finaldir_);
+        }
+    }
+    catch (std::exception const& e)
+    {
+        JLOG(j_.error()) << "deterministic shard " << index_
+                         << " failed to close due to exception: " << e.what();
+    }
+}
+
+void
+DeterministicShard::store(std::shared_ptr<NodeObject> const& nObj)
+{
+    try
+    {
+        backend_->store(nObj);
+        ++curMemObjs_;
+    }
+    catch (std::exception const& e)
+    {
+        JLOG(j_.error()) << "deterministic shard " << index_
+                         << " failed to store due to exception: " << e.what();
+    }
+}
+
+void
+DeterministicShard::flush(bool final)
```

Comment:
Here is an alternate solution you might want to use. It removes the public `flush` function and has `store` automatically flush. https://github.com/miguelportilla/rippled/commit/62942f67848419e241911bdce39b28916947b91b

Reply:
Done.